fpmsyncd/bgpmon.py

```diff
@@ -0,0 +1,171 @@
+    previous get request but no longer there in the current get request), a
+    "previous" neighbor dictionary will be kept and used to determine if there
+    is a need to perform update or the peer is stale to be removed from the
+    state DB
+"""
+
+import commands
+import json
+import os
+import syslog
+import swsssdk
+import time
+import traceback
+
+class BgpStateGet():
+    def __init__(self):
+        self.ipv4_n = []
+        self.ipv4_n_state = {}
+        self.ipv6_n = []
+        self.ipv6_n_state = {}
+        self.new_ipv4_n = []
+        self.new_ipv4_n_state = {}
+        self.new_ipv6_n = []
+        self.new_ipv6_n_state = {}
+        self.retrieve_fail_cnt = 0
+        self.cached_timestamp = 0
+        self.db = swsssdk.SonicV2Connector(host='127.0.0.1')
+        self.db.connect(self.db.STATE_DB, False)
+        self.db.delete_all_by_pattern(self.db.STATE_DB, "NEIGH_STATE_TABLE|*" )
+
+    # A quick way to check if there are anything happening within BGP is to
+    # check its log file has any activities. This is by checking its modified
+    # timestamp against the cached timestamp that we keep and if there is a
+    # difference, there is activity detected. In case the log file got wiped
+    # out, it will default back to constant pulling every 15 seconds
+    def bgp_activity_detected(self):
+        try:
+            timestamp = os.stat("/var/log/frr/frr.log").st_mtime
+            if timestamp != self.cached_timestamp:
+                #syslog.syslog("bgpmon timestamp diff detected: {}".format(timestamp))
+                self.cached_timestamp = timestamp
+                return True
+            else:
+                return False
+        except Exception, e:
+            return True
+
+    # Get a new snapshot of BGP neighbors and store them in the "new" location
+    def get_all_neigh_states(self):
+        # Clean up the "new" dictionaries
+        del self.new_ipv4_n[:]
+        self.new_ipv4_n_state.clear()
+        del self.new_ipv6_n[:]
+        self.new_ipv6_n_state.clear()
+        try:
+            cmd = "vtysh -c 'show bgp summary json'"
+            output = commands.getoutput(cmd)
+            peer_info = json.loads(output)
+            if "ipv4Unicast" in peer_info and "peers" in peer_info["ipv4Unicast"]:
+                self.new_ipv4_n = peer_info["ipv4Unicast"]["peers"].keys()
+                for i in range (0, len(self.new_ipv4_n)):
+                    self.new_ipv4_n_state[self.new_ipv4_n[i]] = \
+                    peer_info["ipv4Unicast"]["peers"][self.new_ipv4_n[i]]["state"]
+
+            if "ipv6Unicast" in peer_info and "peers" in peer_info["ipv6Unicast"]:
+                self.new_ipv6_n = peer_info["ipv6Unicast"]["peers"].keys()
+                for i in range (0, len(self.new_ipv6_n)):
+                    self.new_ipv6_n_state[self.new_ipv6_n[i]] = \
+                    peer_info["ipv6Unicast"]["peers"][self.new_ipv6_n[i]]["state"]
```

Comment:
ipv4 and ipv6 code are quite similar, could you reuse code?

Reply:
@qiluo-msft I think readability is important and I do want to keep IPV4 and IPV6 in separate DS so that we have a choice to handle them independently/differently if needed in future.  The lines involved currently is very small already so Let's keep it this way for now until later when we add more feature to this daemon and find the separation becomes unnecessary and can then combine them at that time.

Reply:
@qiluo-msft After giving some more thoughts, I agree with your comment.  Since there are no clear differences in the handling currently between IPV4 and IPV6 I will keep all the peers and peer state in the same list and dictionary thus removing the duplication of code handling.  In my next check in this should be reflected.